crates/holochain_persistence_api/src/txn/txn.rs

```diff
@@ -0,0 +1,133 @@
+/// Transactional trait extensions to the CAS and EAV persistence
+use crate::{cas::storage::ContentAddressableStorage, eav::*, error::*};
+use std::marker::PhantomData;
+
+/// Defines a transactional writer, typically implemented over a cursor.
+pub trait Writer {
+    /// Commits the transaction. Returns a `PersistenceError` if the
+    /// transaction does not succeed.
+    fn commit(self) -> PersistenceResult<()>;
+}
+
+/// Cursor interface over both CAS and EAV databases. Provides transactional support
+/// by providing a `Writer` across both of them.
+pub trait Cursor<A: Attribute>:
+    Writer + ContentAddressableStorage + EntityAttributeValueStorage<A>
+{
+}
+
+// TODO Should cursor's even be cloneable? SPIKE this
+clone_trait_object!(<A:Attribute> Cursor<A>);
+
+/// A write that does nothing, for testing or for
+/// impementations that don't require the commit and abort functions
+/// to do anything critical.
+#[derive(Clone)]
+pub struct NoopWriter;
+
+impl NoopWriter {
+    pub fn new() -> Self {
+        NoopWriter
+    }
+}
+
+impl Writer for NoopWriter {
+    fn commit(self) -> PersistenceResult<()> {
+        Ok(())
+    }
+}
+
+/// Creates cursors over both EAV and CAS instances. May acquire read or write
+/// resources to do so, depending on implementation.
+///
+/// Some cursors may cascade over temporary (aka "scratch") databases to improve
+/// concurrency performance.
+///
+/// An advanced cursor might wrap other cursors and check external resources (such as peer data in a network)
+/// in addition to the cursors it wraps.
+pub trait CursorProvider<A: Attribute> {
+    /// The type of a cursor for this cursor provider
+    type Cursor: Cursor<A>;
+
+    /// Creates a new cursor. Use carefully as one instance of a cursor may block another,
+    /// especially when cursors are mutating the primary store.
+    fn create_cursor(&self) -> PersistenceResult<Self::Cursor>;
+}
+
+/// A high level api which brings together a CAS, EAV, and
+/// Cursor over them both. A cursor may start transactions over both
+/// the stores or not, depending on implementation.
+pub trait PersistenceManager<A: Attribute>: CursorProvider<A> {
+    /// The type of Content Addressable Storage (CAS)
+    type Cas: ContentAddressableStorage;
+    /// The type of Entity Entity AttributeValue Storage (EAV)
+    type Eav: EntityAttributeValueStorage<A>;
+
+    /// Gets the CAS storage.
+    fn cas(&self) -> Self::Cas;
+    /// Gets the EAV storage
+    fn eav(&self) -> Self::Eav;
+}
```

Comment:
The clone traits are part of the persistence api already- we can open a separate issue to clean this up though. The holochain-rust repo depends on these to be cloneable quite pervasively at the moment. 

The reason why the EAV and CAS are here: 
  - holochain-rust probably still needs references to them without cursors in some places for legacy purposes
  - it provides a unified interface to get both the CAS and EAV since they may share resources such as transactions. In other words, its a high level api for all persistence related stuff (including cursors).

It may be that the associated type approach is wrong but we do need, at least initially in holochain-rust, to share CAS the EAV within several structures. I was considering making a version of the persistence manager which only requires an attribute type and otherwise uses dynamic traits for the CAS, EAV, cursor, etc. This will at least avoid the explosion of type parameters for users of the api.

Reply:
I have no problem with having a unified interface. Let me show you the two alternatives I have in mind for defining this trait. (I'm not sure which is best of the three)

Current
```rust
pub trait PersistenceManager<A: Attribute>: CursorProvider<A> {
    /// The type of Content Addressable Storage (CAS)
    type Cas: ContentAddressableStorage;
    /// The type of Entity Entity AttributeValue Storage (EAV)
    type Eav: EntityAttributeValueStorage<A>;

    /// Gets the CAS storage.
    fn cas(&self) -> Self::Cas;
    /// Gets the EAV storage
    fn eav(&self) -> Self::Eav;
}
```
Super-trait
```rust
pub trait PersistenceManager<A: Attribute>: 
    CursorProvider<A>
   + ContentAddressableStorage
    + EntityAttributeValueStorage<A>
{}
```

Returning references
```rust
pub trait PersistenceManager<A: Attribute>: CursorProvider<A> {
    /// The type of Content Addressable Storage (CAS)
    type Cas: ContentAddressableStorage;
    /// The type of Entity Entity AttributeValue Storage (EAV)
    type Eav: EntityAttributeValueStorage<A>;

    /// Gets the CAS storage.
    // note: returning by reference
    //               v
    fn cas(&self) -> &Self::Cas;
    /// Gets the EAV storage
    // note: returning by reference
    //               v
    fn eav(&self) -> &Self::Eav;
}
```